### PR TITLE
Smooth Sampling / Quadratic Sampling support

### PR DESCRIPTION
**Pull request by kalomaze**
# Smooth Sampling / Quadratic Sampling
This sampling method differs from the truncation samplers (Top K, Top P, Min P) or traditional Temperature sampling, since we are changing the raw scores in a non-linear fashion.

image

With this approach, we tweak the original logit scores using a quadratic transformation, based on each score's distance from the topmost logit.

You can view this as an alternative to Temperature that scales differently (though they are not mutually exclusive); the idea is that we can make the good choices more evenly probable while still punishing extreme outlier tokens.

Graph demonstration with voiceover (excuse how I sound here): [https://files.catbox.moe/x804ia.mp4](https://files.catbox.moe/x804ia.mp4)

This has been implemented in koboldcpp and text-generation-webui for a while now already, and it recieved some praise there.
Considering that, I wanted to backport it to llama.cpp for use with the server.

## How is this meaningfully different from Temperature?
The interesting element is that even the higher, more deterministic values will avoid biasing towards the topmost token if there's a group of similar probability tokens at the top, which meaningfully differs from lower Temperature having a linear skew.
So instead of the top two tokens that were originally 51/49 becoming more skewed towards 60/40 with a temperature of say, 0.3, it would look more like a 50/50 dead even split with a high "smoothing factor" value, while still nullifying the rest of the distribution's low probability outliers.
Likewise, low values will make the top probabilities less deterministic and are a decent fit for creative writing, as low probability outliers are still being reduced "smoothly" without any specific cutoffs.

## How do I scale it?
"0" turns off the sampler entirely. 0.01 would be extremely close to a flat distribution (so extremely unhinged like higher Temperature would be).
You can, in theory, scale to arbritarily large values from here, where it will become gradually more and more deterministic.
Consider 10.0 as a reasonable "max", but that's just an arbitrary limit.
"Smoothing Factor" values of 0.2-0.3 are generally good for creative writing.

## Preview
Here is a useful webpage that allows you to visualize how the distribution changes in response to this sampler:
https://artefact2.github.io/llm-sampling/index.xhtml
Here is the text-generation-webui PR where I first implemented this: https://github.com/oobabooga/text-generation-webui/pull/5403

### CODE REVIEW

**Comment by kalomaze** [comment]
I would also like to point out that Dynamic Temperature is unrelated from this, and it is not required to use the quadratic transformation, they are just combined in one "entropy" function for the sake of not creating an additional function.

**Comment by kalomaze** [comment]
Backported the "curve", which is the 2nd hyperparameter.
This adds a cubic transformation on top of the quadratic one, and can help make lower smoothing_factor values work if the curve is set higher.
1.0 = equivalent to no change from just the quadratic transformation described earlier.

image

https://github.com/oobabooga/text-generation-webui/pull/5551

**Comment by MaggotHATE** [comment]
```json
"temp": 1.5,
"dynatemp_range": 1.4,
"smoothing_factor": 0.2,
"smoothing_curve": 1.5,
"samplers_sequence": "kt",
```

And it just works, no major problems with repetition (even though penalties are switched off).

One question: dynatemp (and smoothing now) are not currently used with Mirostat. Would it be beneficial to use them together, theoretically speaking? I tried it once, but Mirostat was difficult to work with originally, and is difficult to compare now.

**Review by ggerganov — APPROVED** [review]

**Comment by kalomaze** [comment]
> I know it's late to change it, but I really don't like how there's a huge discontinuity in behaviour between 0.0 and small positive values. None of the other samplers are like this, and I expect it will cause confusion to a lot of users that are mentally used to "it's close to the off value, so it means small effects right?".

I can't think of a mathematically identical way to scale the logits that wouldn't require bounding to some arbritary value / taking away control from the user unfortunately.

> It would be nice to have a separate `llama_sample_smoothing` so it can be used on its own, independent of entropy/dynatemp. It looks like you can get the same effect here by calling `llama_sample_entropy(ctx, candidates, 1, 1, 1, smoothing_factor, smoothing_curve)` so the entropy part has no effect, but that seems a bit silly, and also does a bunch of unneeded work (the entropy part doesn't have a fast path for the case where its arguments make it into a no-op)

I imagine @ggerganov would probably prefer if I did this before it can be merged, since the entropy DynaTemp is distinct from this technique. Though, it would probably require a new case switch / new position in the sampler order if we want temperature + this to be "stackable". How does that sound?

**Comment by ggerganov** [comment]
In my opinion, we already have more than enough samplers implemented. The interface allows user code to implement custom sampling techniques if more is needed.

Still, if we want to extend further, we can do it. But we should write some unit tests also. Maybe add some tests for this functionality and merge it for now?

**Comment by jukofyork** [comment]
Have you tried using `x*Exp[k*x]`?

It's intrinsically linked to the softmax function and the only scaling function which is invariant to translation in the same way as the logits are (meaning you don't need to subtract from the maximum logit). 

If you reparameterise the temperature as `Exp[k*1] = Exp[k]`, then look at `Exp[k*x]` and `Exp[k*x^2]` and so on, then the relationship to polynomials should be clearer:

- Standard temperature scaling is akin to the constant term.
- `Exp[k*x]` is akin the the linear term where ratios are invariant (ie: linear).
- `Exp[k*x^2]`is akin to the quadratic term and now isn't invariant (ie: nonlinear), and should likely be 'centered' on the maximum logit value in the same way as you did with your example.

If you do 'center' them then interestingly `Exp[k*x^2]` is an unnormalised Normal distribution and `Exp[k*x]` is an unnormalised Laplace distribution.

So the general transformation is: `x*Exp[a]*Exp[b*x]*Exp[c*x^2]... = x*Exp[a + bx + cx^2 + ...] ` which again shows the similarly to polynomial transformation.

**Comment by mofosyne** [comment]
Hows this PR? Is there a positive intent now to start merging this?

**Comment by SilverLatios** [comment]
Hi, now that DRY was merged, is it possible to continue working on this pr?

**Comment by KeinNiemand** [comment]
Any chance this PR could still move forward, or is there a preferred path for getting smoothing/quadratic sampling into llama.cpp now? I’d be curious whether this PR could be revived or if a fresh PR would be more welcome.
